src/Functions/FunctionBinaryArithmetic.h

```diff
@@ -1125,6 +1140,86 @@ class FunctionBinaryArithmetic : public IFunction
     }
 
+    template <typename ColumnType>
+    ColumnPtr executeArrayPlusMinus(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type,
+                                    size_t input_rows_count, const FunctionOverloadResolverPtr & function_builder) const
+    {
+        auto function = function_builder->build(arguments);
+        return function->execute(arguments, result_type, input_rows_count);
+    }
+
+    static ColumnPtr callFunctionNotEquals(ColumnWithTypeAndName first, ColumnWithTypeAndName second, ContextPtr context)
+    {
+        ColumnsWithTypeAndName args{first, second};
+        auto eq_func = FunctionFactory::instance().get("notEquals", context)->build(args);
+        return eq_func->execute(args, eq_func->getResultType(), args.front().column->size());
+    }
+
+    ColumnPtr executeArrayImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const
+    {
+        const auto * return_type_array = checkAndGetDataType<DataTypeArray>(result_type.get());
+
+        if (!return_type_array)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Return type for function {} must be array.", getName());
+
+        if constexpr (is_multiply || is_division)
+            throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "Cannot use multiplication or division on arrays");
+
+        auto num_args = arguments.size();
+        DataTypes data_types;
+
+        ColumnsWithTypeAndName new_arguments {num_args};
+        DataTypePtr result_array_type;
+
+        const auto * left_const = typeid_cast<const ColumnConst *>(arguments[0].column.get());
+        const auto * right_const = typeid_cast<const ColumnConst *>(arguments[1].column.get());
+
+        /// Unpacking arrays if both are constants.
+        if (left_const && right_const)
+        {
+            new_arguments[0] = {left_const->getDataColumnPtr(), arguments[0].type, arguments[0].name};
+            new_arguments[1] = {right_const->getDataColumnPtr(), arguments[1].type, arguments[1].name};
+            auto col = executeImpl(new_arguments, result_type, 1);
+            return ColumnConst::create(std::move(col), input_rows_count);
+        }
+
+        /// Unpacking arrays if at least one column is constant.
+        if (left_const || right_const)
+        {
+            new_arguments[0] = {arguments[0].column->convertToFullColumnIfConst(), arguments[0].type, arguments[0].name};
+            new_arguments[1] = {arguments[1].column->convertToFullColumnIfConst(), arguments[1].type, arguments[1].name};
+            return executeImpl(new_arguments, result_type, input_rows_count);
+        }
+
+        /// Unpacking non-const arrays and checking sizes of them.
+        if (*typeid_cast<const ColumnArray *>(arguments[0].column.get())->getOffsets().data() !=
+            *typeid_cast<const ColumnArray *>(arguments[1].column.get())->getOffsets().data())
```

Comment:
Shouldn't we compare all of the offsets? Right now we only compare the first offsets, but what if the second arrays are not equally sized?

Reply:
Example:
```
ip-10-2-3-206.eu-west-1.compute.internal :) CREATE TABLE a ( x Array(UInt64), y Array(UInt64)) ENGINE = Memory;

CREATE TABLE a
(
    `x` Array(UInt64),
    `y` Array(UInt64)
)
ENGINE = Memory

Query id: 4ecb0937-c037-47a5-87ab-0cc63365661b

Ok.

0 rows in set. Elapsed: 0.009 sec. 

ip-10-2-3-206.eu-west-1.compute.internal :) INSERT INTO a VALUES ([2,3],[4,5]),([1,2,3], [4,5]),([6,7],[8,9,10])

INSERT INTO a FORMAT Values

Query id: ba4b7a75-db98-4cb9-8670-084bb9409f0e

Ok.

3 rows in set. Elapsed: 0.011 sec. 

ip-10-2-3-206.eu-west-1.compute.internal :) SELECT * FROM a;

SELECT *
FROM a

Query id: d621d886-87b3-4202-a573-ae045448d5fa

┌─x───────┬─y────────┐
│ [2,3]   │ [4,5]    │
│ [1,2,3] │ [4,5]    │
│ [6,7]   │ [8,9,10] │
└─────────┴──────────┘

3 rows in set. Elapsed: 0.007 sec. 

ip-10-2-3-206.eu-west-1.compute.internal :) SELECT x, y, x+y FROM a;

SELECT
    x,
    y,
    x + y
FROM a

Query id: 303c7af3-c518-4977-975a-bccc01267da3

┌─x───────┬─y────────┬─plus(x, y)─┐
│ [2,3]   │ [4,5]    │ [6,8]      │
│ [1,2,3] │ [4,5]    │ [5,7,11]   │
│ [6,7]   │ [8,9,10] │ [15,17]    │
└─────────┴──────────┴────────────┘

3 rows in set. Elapsed: 0.009 sec. 
```

Here the x+y is not correct, because the second and third row don't have matching sized arrays in column `x` and `y`.